### DISCUSSION TITLE
How to create a Job that can inject dependencies using DI ?

### DISCUSSION THREAD

**Discussion by amirdar4** [quartznet/quartznet#2512] [Q&A]
Hi all

I got a Job class that should invoke some function on another class.

The job class looks like this:

```
using NanoLockSec.Services.Identity.API.Managers.Interfaces;
using Quartz;

namespace NanoLockSec.Services.Identity.API.Jobs
{
    [DisallowConcurrentExecution]
    public class ActiveDirectoryJob : IJob
    {
        private readonly ILogger<ActiveDirectoryJob> _logger;
        private readonly IActiveDirectoryIntegrator _activeDirectoryIntegrator;

        public ActiveDirectoryJob(
            ILogger<ActiveDirectoryJob> logger,
            IActiveDirectoryIntegrator activeDirectoryIntegrator)
        {
            _logger = logger;
            _activeDirectoryIntegrator = activeDirectoryIntegrator;
        }

        public async Task Execute(IJobExecutionContext context)
        {
            _logger.LogInformation("Executing Active Directory job");
            var searchResponse = await _activeDirectoryIntegrator.GetActiveDirectoryUsers();
            _logger.LogInformation("Active Directory job ended");
        }
    }
}
```
in my Program.cs i added:

```
builder.Services.AddTransient<ActiveDirectoryJob>();
builder.Services.AddQuartz(q =>
{
    q.UseMicrosoftDependencyInjectionJobFactory();
    q.SchedulerId = "Quartz-Scheduler";
    q.SchedulerName = "Quartz.NET Scheduler";
    q.UseMicrosoftDependencyInjectionJobFactory();
});

builder.Services.AddQuartzServer(options =>
 {
   options.WaitForJobsToComplete = true;
}); 
```

and I created the scheduler for the job like this (this is called by the user with some button for the simplicity of the flow)

```
private async Task<bool> CreateScheduler()
{
	IJobDetail job = JobBuilder.Create<ActiveDirectoryJob>().WithIdentity("job1", "group1").Build();

	ITrigger trigger = TriggerBuilder.Create()
		.WithIdentity("trigger1", "group1")
		.StartNow()
		.WithSimpleSchedule(x => x.WithIntervalInSeconds(10).RepeatForever())
		.Build();

	var scheduler = await StdSchedulerFactory.GetDefaultScheduler();
	await scheduler.ScheduleJob(job, trigger);
	await scheduler.Start();
	
	return scheduler.IsStarted;
}
```
while scheduler.IsStarted returns "True" I get an error that says:

An error occurred instantiating job to be executed. job= 'group1.job1'
      Quartz.SchedulerException: Problem instantiating class 'NanoLockSec.Services.Identity.API.Jobs.ActiveDirectoryJob: Cannot instantiate type which has no empty constructor (Parameter 'ActiveDirectoryJob')'

i looked around for a solution, and all I found was that I should add the "UseMicrosoftDependencyInjectionJobFactory()" - but this is now deprecated, and also not working even if I use it.

see: [sof](https://stackoverflow.com/questions/55333351/quartz-schedulerexception-problem-instantiating-class-quartz-is-not-executing)

so, I'm stuck on something that should be pretty basic IMO.

Any help would be highly appreciated

**Comment by lahma** [comment]
Quite similar to this: https://stackoverflow.com/questions/78788778/quartz-dependency-injection-with-net8

Don't create separate scheduler factory.

**Reply by amirdar4** [reply]
hi
thanks for the quick reply.
I've gone over this link, and all the documentation and i still can't get how to inject my object into the Job class.

I'm having a basic misunderstanding of how Quartz should work.
from what I understand there are 3 key components: the Scheduler, the Job, and the Trigger
the Scheduler should invoke the Job "Execute" function based on the Trigger rules.
so far, it's all good (i hope)

what confuses me is that all the initialization of jobs, triggers, and schedulers always happens inside the "AddQuarz" extension method. (in Program.cs)

in my scenario, the Job should run every time a user saves some configuration, i.e., press a button.
so it can't be done as part of the "AddQuarz" extension method since this only occurs while the program starts running.
**I need a way to start\stop\delete the scheduler while I save a configuration (again, some button press) in run time**

Hope this clarify the problem

**Reply by lahma** [reply]
Do the minimal in `AddQuartz()` and then obtain reference to `ISchedulerFactory` from DI container and call it's `GetScheduler()` to get the scheduler reference. You can use the `IScheduler` API to start/stop and add/remove jobs and trigggers.

**Reply by amirdar4** [reply]
@lahma  that worked! I'm still not sure why
`var scheduler = await schedulerFactory.GetScheduler(); `
is working, while 
`var scheduler = await StdSchedulerFactory.GetDefaultScheduler();`
is not

also, I have another question if that is okay with you

as I said, I need a way to stop\start\update my scheduler based on a user choice
I'm wondering what is the best practice for this:

currently, I'm doing this:
*for a new scheduler:    
`await scheduler.ScheduleJob(job, trigger);`
*when need to update the interval:
`  await scheduler.RescheduleJob(triggerKey, newTrigger);`
(newTrigger will have the new interval)
* if the user decides to stop the scheduler:
`await scheduler.Shutdown();`

I'm not sure if I should Shutdown the schedule or maybe do something like:  
`pauseJob(JobKey jobKey) `
and 
`resumeJob(JobKey jobKey)`

thanks !
